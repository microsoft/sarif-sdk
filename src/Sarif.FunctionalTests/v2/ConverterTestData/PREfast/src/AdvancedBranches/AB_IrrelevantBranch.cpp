// Make a branch irrelevant if an error occurs on both sides

int IrrelevantBranch_6011(bool flag, int input)
{
	int *source1 = 0, *source2 = 0;

	if (flag) {
		source2 = &input;
	}

	int *target = source1;
	return *target;
}

// This should work even if there are multiple defects 
// interacting with the branch

void C6385_MaybeRelevantBranch(int flag)
{
	int data[10];
	int index = 10;

	if (flag<10) {
		index = 9;
	}

	int k = data[index];
}

