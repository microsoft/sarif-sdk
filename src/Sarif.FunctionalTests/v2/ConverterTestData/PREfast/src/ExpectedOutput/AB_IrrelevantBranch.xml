<?xml version='1.0' encoding='UTF-8'?>
<DEFECTS><DEFECT><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>12</LINE><COLUMN>1</COLUMN></SFA><DEFECTCODE>6011</DEFECTCODE><DESCRIPTION>Dereferencing NULL pointer 'target'. </DESCRIPTION><FUNCTION>IrrelevantBranch_6011</FUNCTION><DECORATED>?IrrelevantBranch_6011@@YAH_NH@Z</DECORATED><FUNCLINE>3</FUNCLINE><PROBABILITY>1</PROBABILITY><RANK>4</RANK><CATEGORY><RULECATEGORY>mspft</RULECATEGORY></CATEGORY><PATH><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>5</LINE><COLUMN>6</COLUMN><KEYEVENT><ID>1</ID><KIND>declaration</KIND><IMPORTANCE>Essential</IMPORTANCE><MESSAGE>'source1' is NULL</MESSAGE></KEYEVENT></SFA><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>5</LINE><COLUMN>20</COLUMN></SFA><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>7</LINE><COLUMN>5</COLUMN></SFA><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>11</LINE><COLUMN>6</COLUMN><KEYEVENT><ID>2</ID><KIND>alias</KIND><IMPORTANCE>Essential</IMPORTANCE><MESSAGE>(alias) 'target' receives the value of 'source1' (which may be NULL)</MESSAGE></KEYEVENT></SFA><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>12</LINE><COLUMN>1</COLUMN><KEYEVENT><ID>3</ID><KIND>usage</KIND><IMPORTANCE>Essential</IMPORTANCE><MESSAGE>'target' is dereferenced, but may still be NULL</MESSAGE></KEYEVENT></SFA></PATH></DEFECT><DEFECT><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>27</LINE><COLUMN>5</COLUMN></SFA><DEFECTCODE>6001</DEFECTCODE><DESCRIPTION>Using uninitialized memory 'data'.</DESCRIPTION><FUNCTION>C6385_MaybeRelevantBranch</FUNCTION><DECORATED>?C6385_MaybeRelevantBranch@@YAXH@Z</DECORATED><FUNCLINE>18</FUNCLINE><PROBABILITY>1</PROBABILITY><RANK>3</RANK><CATEGORY><RULECATEGORY>mspft</RULECATEGORY></CATEGORY><PATH><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>20</LINE><COLUMN>5</COLUMN><KEYEVENT><ID>1</ID><KIND>declaration</KIND><IMPORTANCE>Essential</IMPORTANCE><MESSAGE>'data' is not initialized</MESSAGE></KEYEVENT></SFA><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>21</LINE><COLUMN>5</COLUMN></SFA><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>23</LINE><COLUMN>9</COLUMN></SFA><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>27</LINE><COLUMN>5</COLUMN><KEYEVENT><ID>2</ID><KIND>usage</KIND><IMPORTANCE>Essential</IMPORTANCE><MESSAGE>'data' is used, but may not have been initialized</MESSAGE></KEYEVENT></SFA></PATH></DEFECT><DEFECT><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>27</LINE><COLUMN>5</COLUMN></SFA><DEFECTCODE>6385</DEFECTCODE><DESCRIPTION>Reading invalid data from 'data':  the readable size is '40' bytes, but '44' bytes may be read.</DESCRIPTION><FUNCTION>C6385_MaybeRelevantBranch</FUNCTION><DECORATED>?C6385_MaybeRelevantBranch@@YAXH@Z</DECORATED><FUNCLINE>18</FUNCLINE><PROBABILITY>1</PROBABILITY><RANK>4</RANK><CATEGORY><RULECATEGORY>mspft</RULECATEGORY></CATEGORY><PATH><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>20</LINE><COLUMN>5</COLUMN><KEYEVENT><ID>1</ID><KIND>declaration</KIND><IMPORTANCE>Essential</IMPORTANCE><MESSAGE>'data' is an array of 10 elements (40 bytes)</MESSAGE></KEYEVENT></SFA><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>21</LINE><COLUMN>5</COLUMN><KEYEVENT><ID>2</ID><KIND>declaration</KIND><IMPORTANCE>Essential</IMPORTANCE><MESSAGE>'index' is equal to 10</MESSAGE></KEYEVENT></SFA><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>23</LINE><COLUMN>9</COLUMN><KEYEVENT><ID>3</ID><KIND>branch</KIND><IMPORTANCE>Full</IMPORTANCE><MESSAGE>Skip this branch, (assume 'flag&lt;10' is false)</MESSAGE></KEYEVENT></SFA><SFA><FILEPATH>e:\enlistcsa\sources\wincsa\src\ddsuites\src\vs\edtqa\static\xmlvalidation\advancedbranches\</FILEPATH><FILENAME>ab_irrelevantbranch.cpp</FILENAME><LINE>27</LINE><COLUMN>5</COLUMN><KEYEVENT><ID>4</ID><KIND>usage</KIND><IMPORTANCE>Essential</IMPORTANCE><MESSAGE>Invalid read from 'data[10]', (readable range is 0 to 9)</MESSAGE></KEYEVENT></SFA></PATH></DEFECT></DEFECTS>